torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -388,8 +410,32 @@ void ProcessGroupAgent::handleSend(const SendWork& work) {
     pendingSends.emplace_back(pg_->send(preamble, dst, dst /* channelTag */));
     pendingSends.emplace_back(pg_->send(payload, dst, dst /* channelTag */));
   }
+  // Write pendingSends to a global map so that they can be interrupted by
+  // ::shutdown().
+  std::unique_lock<std::mutex> guard(pendingSendMutex_);
+
+  for (auto& p : pendingSends) {
+    currentPendingSends_[dst].push_back(p);
+  }
+
+  // Unlock to call into wait, otherwise shutdown() cannot interrupt here.
+  guard.unlock();
+
   for (auto& pendingSend : pendingSends) {
-    pendingSend->wait();
+    if (!rpcRunning_.load() || !pendingSend->wait()) {
+      // Send was interrupted or RPC is not running.
+      return;
+    }
+  }
+
+  // re-lock to erase
+  guard.lock();
+  // NB: We cannot just erase all of currentPendingSends[dst], since this might
+  // preemptively remove sends from other threads.
+  auto& vec = currentPendingSends_[dst];
+  for (auto& p : pendingSends) {
+    auto it = std::find(vec.begin(), vec.end(), p);
```

Comment:
This is linear complexity? Will it be more efficient if we use the thread id to key pending sends in `currentPendingSends_`? Or is there any reason that they have to be keyed using `dst`?

Reply:
I think we can key by the thread_id and then remove all sends corresponding to the thread_id here instead. Alternatively, it looks like we could also turn the vector into a set which would give us constant lookup since `pendingSends` here is only always of size 2.